include/circt/Dialect/FIRRTL/Namespace.h

```diff
@@ -59,29 +46,72 @@ class CircuitNamespace {
   /// 1. The original name is returned.
   /// 2. The name is given a `_<n>` suffix where `<n>` is a number starting from
   ///    `_0` and incrementing by one each time.
-  std::string newName(const Twine &name) {
+  StringRef newName(const Twine &name) {
     // Special case the situation where there is no name collision to avoid
     // messing with the SmallString allocation below.
     llvm::SmallString<64> tryName;
-    StringRef workingName = name.toStringRef(tryName);
-    if (internal.insert(workingName).second)
-      return workingName.str();
+    auto inserted = internal.insert(name.toStringRef(tryName));
+    if (inserted.second)
+      return inserted.first->getKey();
 
     // Try different suffixes until we get a collision-free one.
     size_t i = 0;
     if (tryName.empty())
       name.toVector(tryName); // toStringRef may leave tryName unfilled
     tryName.push_back('_');
     size_t baseLength = tryName.size();
-    do {
+    for (;;) {
       tryName.resize(baseLength);
       Twine(i++).toVector(tryName); // append integer to tryName
-    } while (!internal.insert(tryName).second);
-    return std::string(tryName);
+      auto inserted = internal.insert(tryName);
+      if (inserted.second)
+        return inserted.first->getKey();
+    }
+  }
+
+protected:
+  llvm::StringSet<> internal;
+};
+
+/// The namespace of a `CircuitOp`, generally inhabited by modules.
+class CircuitNamespace : public NamespaceBase {
+public:
```

Comment:
You would know better than me, but should this just be a `struct` given that all members are public?  Same for `ModuleNamespace`?

Reply:
Yeah true, I don't see a reason why this shouldn't be a struct either. AFAIK in clang the only difference is just the default `public:` visibility, but there was something about MSVC being sensitive to the difference.